core/commands/dag/dag.go

```diff
@@ -241,3 +278,421 @@ var DagResolveCmd = &cmds.Command{
 	},
 	Type: ResolveOutput{},
 }
+
+type expectedRootsSeen map[cid.Cid]bool
+type importResult struct {
+	roots *expectedRootsSeen
+	err   error
+}
+
+var DagImportCmd = &cmds.Command{
+	Helptext: cmds.HelpText{
+		Tagline: "Import the contents of .car files",
+		ShortDescription: `
+'ipfs dag import' parses .car files and adds all objects to the blockstore.
+By default, after all car files have been processed, an attempt is made to
+pin each individual root specified in the car headers, before GC runs again.
+`,
+	},
+	Arguments: []cmds.Argument{
+		cmds.FileArg("path", true, true, "The path of a .car file.").EnableStdin(),
+	},
+	Options: []cmds.Option{
+		cmds.BoolOption(progressOptionName, "p", "Display progress on CLI.").WithDefault(true),
+		cmds.BoolOption(silentOptionName, "No output."),
+		cmds.BoolOption(pinRootsOptionName, "Pin optional roots listed in the .car headers after importing.").WithDefault(true),
+	},
+	PreRun: func(req *cmds.Request, env cmds.Environment) error {
+		if silent, _ := req.Options[silentOptionName].(bool); silent {
+			req.Options[progressOptionName] = false
+		}
+
+		return nil
+	},
+	Type: CarImportOutput{},
+	Run: func(req *cmds.Request, res cmds.ResponseEmitter, env cmds.Environment) error {
+
+		node, err := cmdenv.GetNode(env)
+		if err != nil {
+			return err
+		}
+
+		api, err := cmdenv.GetApi(env, req)
+		if err != nil {
+			return err
+		}
+
+		// on import ensure we do not reach out to the network for any reason
+		// if a pin based on what is imported + what is in the lockstore
+		// isn't possible: tough luck
+		api, _ = api.WithOptions(options.Api.Offline(true))
+
+		// grab a gc lock so that regardless of the size of the streamed-in cars
+		// nothing will disappear on us before we had a chance to pin the roots
+		// at the very end
+		// This is especially important for use cases like dagger:
+		//    ipfs dag import $( ... | ipfs-dagger --stdout=carfifos )
+		//
+		// !!!NOTE!!! - We take .GCLock, *not* .PinLock. The way the current stack is
```

Comment:
Take the pin lock. The GCLock is for the GC process.

Reply:
Done: https://github.com/ipfs/go-ipfs/pull/7011/commits/0dae3a2248e6d149193362d997d75a15642fc094

I will open a separate issue for the test in question.

Reply:
( the sharness test gives me sufficient confidence the lock is enough, and we won't lose blocks )